tests/utils_tests/test_lazyobject.py

```diff
@@ -284,3 +287,95 @@ def test_list_set(self):
         self.assertEqual(len(lazy_list), 5)
         self.assertEqual(len(lazy_set), 4)
+
+
+class BaseBaz(object):
+    """
+    A base class with a funky __reduce__ method.
+
+    Meant to simulate the __reduce__ method of Model, which sets
+    self._django_version.
+    """
+    def __init__(self):
+        self.baz = 'wrong'
+
+    def __reduce__(self):
+        self.baz = 'right'
+        return super(BaseBaz, self).__reduce__()
+
+    def __eq__(self, other):
+        if self.__class__ != other.__class__:
+            return False
+        for attr in ['bar', 'baz', 'quux']:
+            if hasattr(self, attr) != hasattr(other, attr):
+                return False
+            elif getattr(self, attr, None) != getattr(other, attr, None):
+                return False
+        return True
+
+
+class Baz(BaseBaz):
+    """
+    A class that inherits from BaseBaz and has its own __reduce_ex__ method.
+    """
+    def __init__(self, bar):
+        self.bar = bar
+        super(Baz, self).__init__()
+
+    def __reduce_ex__(self, proto):
+        self.quux = 'quux'
+        return super(Baz, self).__reduce_ex__(proto)
+
+
+class BazProxy(Baz):
+    """
+    A class that acts as a proxy for Baz.
+
+    It does some scary mucking about with dicts, which simulates some crazy
+    things that people might do with e.g. proxy models.
+    """
+    def __init__(self, baz):
+        self.__dict__ = baz.__dict__
+        self._baz = baz
+        super(BaseBaz, self).__init__()
+
+
+# Regression test for issues with pickling a SimpleLazyObject wrapping a model
+# or other class with a custom __reduce__ method (#25389)
+class SimpleLazyObjectPickleTestCase(TestCase):
+    # Test in a fairly synthetic setting
+    def test_pickle_with_reduce(self):
+        # Test every pickle protocol available
+        for protocol in range(pickle.HIGHEST_PROTOCOL):
+            lazy_objs = [
+                SimpleLazyObject(lambda: BaseBaz()),
+                SimpleLazyObject(lambda: Baz(1)),
+                SimpleLazyObject(lambda: BazProxy(Baz(2))),
+            ]
+            for obj in lazy_objs:
+                pickled = pickle.dumps(obj, protocol)
+                unpickled = pickle.loads(pickled)
+                self.assertEqual(unpickled, obj)
+                self.assertEqual(unpickled.baz, 'right')
+
+    # Test on an actual model, based on the reproduction in #25426
+    def test_pickle_model(self):
+        foo = FooThing.objects.create(name="thing1")
+        BarThing.objects.create(foo=foo)
+
+        # Test every pickle protocol available
+        for protocol in range(pickle.HIGHEST_PROTOCOL):
+            lazy_foo = SimpleLazyObject(lambda: foo)
+            # Test it both if we accessed a field on the model, and if we
+            # didn't.
+            lazy_foo.barthing
+            lazy_foo_2 = SimpleLazyObject(lambda: foo)
+            with warnings.catch_warnings(record=True) as recorded:
+                self.assertEqual(
```

Comment:
single line for these assertions is okay (we prefer longer lines if it helps readability)